src/devices/Ili934x/Ili9341Imaging.cs

```diff
@@ -0,0 +1,167 @@
+        public void DrawBitmap(BitmapImage bm, Rectangle updateRect)
+        {
+            DrawBitmap(bm, new Point(updateRect.X, updateRect.Y), updateRect);
+        }
+
+        /// <summary>
+        /// Send a bitmap to the Ili9341 display specifying the starting position and destination clipping rectangle.
+        /// </summary>
+        /// <param name="bm">The bitmap to be sent to the display controller note that only Pixel Format Format32bppArgb is supported.</param>
+        /// <param name="sourcePoint">A coordinate point in the source bitmap where copying starts from.</param>
+        /// <param name="destinationRect">A rectangle that defines where in the display the bitmap is written. Note that no scaling is done.</param>
+        public void DrawBitmap(BitmapImage bm, Point sourcePoint, Rectangle destinationRect)
+        {
+            if (bm is null)
+            {
+                throw new ArgumentNullException(nameof(bm));
+            }
+
+            FillBackBufferFromImage(bm, sourcePoint, destinationRect);
+        }
+
+        private void FillBackBufferFromImage(BitmapImage image, Point sourcePoint, Rectangle destinationRect)
+        {
+            if (image is null)
+            {
+                throw new ArgumentNullException(nameof(image));
+            }
+
+            Converters.AdjustImageDestination(image, ref sourcePoint, ref destinationRect);
+
+            Parallel.For(destinationRect.Y, destinationRect.Height + destinationRect.Y, y =>
+            {
+                var row = _screenBuffer.AsSpan(y * ScreenWidth, ScreenWidth);
+                for (int i = destinationRect.X; i < destinationRect.Width + destinationRect.X; i++)
+                {
+                    int xSource = sourcePoint.X + i - destinationRect.X;
+                    int ySource = sourcePoint.Y + y - destinationRect.Y;
+                    row[i] = Rgb565.FromRgba32(image[xSource, ySource]);
+                }
+            });
+        }
+
+        /// <summary>
+        /// Updates the display with the current screen buffer.
+        /// <param name="forceFull">Forces a full update, otherwise only changed screen contents are updated</param>
+        /// </summary>
+        public void SendFrame(bool forceFull = false)
+        {
+            if (forceFull)
+            {
+                SetWindow(0, 0, ScreenWidth, ScreenHeight);
+                SendSPI(MemoryMarshal.Cast<Rgb565, byte>(_screenBuffer));
+            }
+            else
+            {
+                int topRow = 0;
+                int bottomRow = ScreenHeight;
+                int w = ScreenWidth;
+                for (int y = 0; y < ScreenHeight; y++)
+                {
+                    for (int x = 0; x < w; x++)
+                    {
+                        if (!Rgb565.AlmostEqual(_screenBuffer[x + y * w], _previousBuffer[x + y * w], 2))
+                        {
+                            topRow = y;
+                            goto reverse;
```

Comment:
consider not using goto in this method

Reply:
I know this is uncommon, but to jump out of nested loops it's really the clearest thing to do IMHO.